beluga_amcl/src/ndt_amcl_node.cpp

```diff
@@ -0,0 +1,915 @@
+
+  return beluga::NDTSensorModel<NDTMapRepresentation>{
+      params, beluga::io::load_from_hdf5_2d<NDTMapRepresentation>(get_parameter("map_path").as_string())};
+}
+
+auto NdtAmclNode::get_execution_policy() const -> ExecutionPolicyVariant {
+  const auto name = get_parameter("execution_policy").as_string();
+  if (name == "seq") {
+    return std::execution::seq;
+  }
+  if (name == "par") {
+    return std::execution::par;
+  }
+  throw std::invalid_argument("Execution policy must be seq or par.");
+};
+
+auto NdtAmclNode::make_particle_filter() const -> std::unique_ptr<NdtAmclVariant> {
+  auto amcl = std::visit(
+      [this](auto motion_model, auto execution_policy) -> NdtAmclVariant {
+        auto params = beluga::AmclParams{};
+        params.update_min_d = get_parameter("update_min_d").as_double();
+        params.update_min_a = get_parameter("update_min_a").as_double();
+        params.resample_interval = static_cast<std::size_t>(get_parameter("resample_interval").as_int());
+        params.selective_resampling = get_parameter("selective_resampling").as_bool();
+        params.min_particles = static_cast<std::size_t>(get_parameter("min_particles").as_int());
+        params.max_particles = static_cast<std::size_t>(get_parameter("max_particles").as_int());
+        params.alpha_slow = get_parameter("recovery_alpha_slow").as_double();
+        params.alpha_fast = get_parameter("recovery_alpha_fast").as_double();
+        params.kld_epsilon = get_parameter("pf_err").as_double();
+        params.kld_z = get_parameter("pf_z").as_double();
+
+        auto hasher = beluga::spatial_hash<Sophus::SE2d>(
+            get_parameter("spatial_resolution_x").as_double(), get_parameter("spatial_resolution_y").as_double(),
+            get_parameter("spatial_resolution_theta").as_double());
+
+        RandomStateGenerator random_state_maker = [](const auto& particles) {
+          static thread_local auto generator = std::mt19937{std::random_device()()};
+          const auto [pose, covariance] =
+              beluga::estimate(beluga::views::states(particles), beluga::views::weights(particles));
+          return [pose, covariance]() { return beluga::MultivariateNormalDistribution{pose, covariance}(generator); };
+        };
+
+        return beluga::Amcl(
+            std::move(motion_model),
+            get_sensor_model(),             //
+            std::move(random_state_maker),  //
+            std::move(hasher),              //
+            params,                         //
+            execution_policy);
+      },
+      get_motion_model(), get_execution_policy());
+  return std::make_unique<NdtAmclVariant>(std::move(amcl));
+}
+
+void NdtAmclNode::timer_callback() {
+  if (!particle_filter_) {
+    return;
+  }
+
+  if (particle_cloud_pub_->get_subscription_count() == 0) {
+    return;
+  }
+
+  // Particle weights from the filter may or may not be representative of the
```

Comment:
@serraramiro1 @glpuga nit: we should relocate this particle cloud computation to `beluga_ros`.

Reply:
I meant the whole computation. We have copies of this in several places.

Reply:
Done.